Logic

r0bd0g opened the pull request:
Bugs Fixes / Other Changes:

1. Kokiri Forest Deku Baba Sticks - I added can_use() for the child item checks. Otherwise, having Kokiri Sword or Boomerang would allow adult to get around the check for shuffle dungeon entrances.

2. Kokiri Forest Deku Baba Nuts - Babas that drop nuts don't actually exist as child, so I removed all of the child items from the check.

3. Zora River Bug Shrub needs some additional logic to reach. You can easily jump there from the lower PoH, so I used the same logic as is used to reach that.

4. Killing the DMT Soil Gold Skulltula with only Strength is too difficult, so I added can_child_attack to it. I didn't think it was worth making a trick for doing it with strength.

5. I reworded the logic on the DC Climb to remove checking for sticks twice (this matches how it was written on the DC Compass Chest). But then I also felt the need to add an additional region to DC to better capture what was going on with the vines GS after adding that trick (see trick 5).

6. There's a bug with the staircase in DC where if you lower it with the Bow trick, then hit the elevator switch, nothing is letting child ride that elevator up so he can use the Boomerang on the above stairs GS. I fixed this by adding an exit from the lobby to the far bridge with logic at(Far Bridge, True), which should allow whatever age didn't hit the switch to be able to go in that direction. This makes the "here"s that are currently in place for DC Climb->Far Bridge irrelevant, so I deleted them.

7. I added the exit from vanilla Forest NW to NE Outdoors, through the well. Also wrapped Forest NW to High Balconies in a here(). Essentially, these changes do nothing. This was done so that child would not have to get a weapon to fight the Bubble in order to get to NE Outdoors to do the Boomerang trick for the Skulltula there, if you dive through the well with Gold Scale or kill the Bubble first as adult. Access to such a weapon however is guaranteed since Boomerang is required to get that Skulltula and Boomerang can be used to get sticks from Babas, so this change only allows the to playthrough not to be forced to pick up an extra weapon.

8. I removed child throwing a bomb at the Skulltula in the first room of Forest Temple from default logic. It took some work but I was able to figure out how to do it consistently. But I feel between the timing and positioning of the bomb throw, there are just too many factors that have to be close to perfect. I sort-of added it back in as a trick, but maybe it was a mistake (see trick 7).

9. I added the "Skulltulas" tag to Fire Temple MQ Flame Wall Maze Skip. (I'm doing real MQ updates in the next round of logic changes.) This trick is only relevant to reach a Skulltula. (For now? But it should keep this tag regardless.)

10. I removed Water Temple Boss Key Chest with Iron Boots. While it is an easier method it's not THAT much easier, and anybody who cares just knows the precise jump. In this same temple is the hover boots vs no hover boots tricks for the cracked wall, but in that case there are clearly people who know some methods but not others.

11. I added a missing route for using FW and Boomerang to get the Water Central Pillar Skulltula. If you don't have Din's or even access to Sticks, but you do have the Water Temple keys (or are on keysy), then you could set FW at the bottom of central pillar as child. Child would have no way of raising the water in this state, so Adult must do it. Bow or Hovers are the logically relevant ways of raising the water. (Hookshot would mean you could just use FW and Hookshot instead of messing around as child.) Of note though, is that the Hovers route would provide no access to the middle water level at any point, but the Skulltula was in the middle water level region. I had to move that Skulltula up to the parent region, which only required additionally checking for Lullaby.

12. I've made significant changes to "Water Temple Dragon Statue with Bombchu". I don't know why I'd never realized this before, but this route for Water Temple Dragon Statue not only allows skipping Irons, but Hookshot as well, so I added the possibility of using Irons instead of a Scale for the trick. That this wasn't accounted for before was just a bug that has now been fixed, so nothing controversial there. However, I've also decided to expand the trick to include additional methods of hitting the switch, now using Bow or Hookshot to hit it from above the water. While the Bombchu and Bow/Hook methods are a bit different and do have their unique challenges, I'd always intended for the Jump Dive variant to be the easy version of the trick, while this trick would be used for the more advanced method. But "method" has now become plural. Both take a bit of practice to understand, but I didn't really feel as if either method was much harder or easier than the other. I feel like the Bombchu method is probably more well-known, but after learning both, I don't really see why it should be so much more popular. The new name for the trick is "Water Temple Dragon Statue Switch from Above the Water as Adult", and with this new name it now makes sense for it to fully supersede the Jump Dive variant. It was always kind of strange before, seeing the logic for going from the serpent river to the dragon statue, and it having the harder variant of the trick require additional items over the easier one. (Now you might ask, why "as Adult"? See trick 10.)

13. The Spirit Temple Nut Crate is a possible first logical nuts to be obtained as Adult (rather than only as child), though Adult will need all Spirit keys (or keysy).


New Tricks:

1. Zora's Domain GS with No Additional Items - A precise jumpslash to kill the Skulltula and recoil back to the top of the frozen waterfall.

2. Goron City Grotto with Hookshot While Taking Damage - You take about 5 hits from the lava floor while doing this, so it can be expected in Double Damage (pretty tough though) but not Quadruple Damage or OHKO.

3. Death Mountain Trail Climb with Hover Boots

4. Hyrule Castle Storms Grotto GS with Just Boomerang - I tried but was unsuccessful at fishing out the token as adult.

5. Dodongo's Cavern Vines GS from Below with Longshot - Skip lowering the staircase by using the Longshot to get it through the vines, which are a one-sided collision.

6. Dodongo's Cavern Two Scrub Room with Strength - Uses a block as a stepping-stone in order to have enough time to make the trek. Only works as Adult.

7. Forest Temple First Room GS with Difficult-to-Use Weapons - Kill it using Sword or Sticks (fall damage required) or child Bomb throws. I wanted to add the trick to kill it with a jumpslash, but it felt weird to do that and also completely remove throwing bombs as child, but I really felt the bomb throws were too difficult for default logic... So I ended up going with a combined trick for both Sword items and child Bombs. It's probably a lapse in judgement. The trick name just looks gross; that's how you can tell.

8. Water Temple Central Pillar GS with Neither Longshot nor Farore's Wind - Open the door, raise the water level, then use the Iron Boots to go back into the room, now with the water at the highest level. On a related note: I think it's time that the Farore's Wind route stopped trolling new players and was removed from default logic. I would have preferred to have combined that logic together into this trick, but after thinking about it, the Farore's Wind route has been a thing for so long now that it at least deserves to be a separate trick. To add it would be as simple as anding in the trick next to both Farore's Wind uses in that section. I wonder though how to handle Farore's Wind potentially being foolish and how people would feel about that. To make changes in this regard would likely require community feedback.

9. Water Temple Falling Platform Room GS with Boomerang - Precise Boomerang throw. Seems like it should have been easy to implement, but it wasn't. To enter the Falling Platform room as child was not supported by the logic, so I had to add that. I had to move region access around a bit as well as add a new region to really capture it. I also had to ensure that if the water is somehow lowered, that you can raise it back up again. For Adult to raise the water, you don't actually even need logical access as Adult since Adult had to have been there to have lowered it in the first place, but in that case you have to check for possible non-repeatable access, which on main branch I believe can occur with shuffled overworld and special interior entrances?

10. Water Temple Dragon Statue Switch from Above the Water as Child - Allows using Slingshot, Boomerang, or Bombchus (though Bombchus are logically irrelevant if the adult version of this trick is enabled) to hit the switch to open the gate, and then an incredibly perfect dive can be used to swim under the gate. The dive has to be perfect because child cannot swim through the gate even while it's open -- he has to dive the whole way under. And of course, the gate will shortly close, so you do not get much time to line up your dive. Gold Scale would make it a bit easier, as being able to dive a bit longer gives you some leniency on actually getting under the gate, but it will all still work with Silver Scale. Silver Scale and Slingshot are a particularly tricky combination, since, with Slingshot, you can't really make much progress toward your destination while the projectile is heading for the switch. I made this a separate trick from the Adult variant mainly because the Child routes are just so much harder (even though the Adult version is already supposed to be an advanced method), due to the requirement that you successfully dive all the way past the gate.

11. Shadow Temple Falling Spikes GS with Hover Boots - This is actually pretty tough.

12. Spirit Temple Main Room Jump from Hands to Upper Ledges - Applies to the Upper Chest and the GS in the main room, to skip Hookshot or Hovers.

13. Ice Cavern Block Room GS with Hover Boots


These are the tricks that I investigated but decided not to add:

1. Death Mountain Trail Bombable Wall GS with Just Boomerang - You had to like, throw the Boomerang while Link's hand is through a wall or something? It seemed more directly glitchy than the other times that tricks can call for Boomerangs to go through walls.

2. Gerudo Fortress Archery GS without Gerudo Card - I didn't like that it only worked because the camera was keeping the guard culled. There's always been a lot of debate over this sort of thing. Cat chain? Boulder collision? So the safe play was to leave this one alone.


Please go over the logic changes with the finest-toothed of combs. This stuff is really hard to get right on the first try.
I welcome questions and comments on the tricks that I both did or didn't decide to add.
And finally please weigh in on the FW situation.
MQ Dungeons logic update coming soon/eventually.

r0bd0g commented:
Toxic pointed out that that water non-repeatable access logic can be foiled by having non-repeatable adult access. (Chus to pass through spirit as child to colossus, temple of time at colossus fairy.) So the check needs to also include shuffle special interiors. This non-repeatable adult access is a wider problem (emptying big poe bottles), but this case is relatively easily fixed.

r0bd0g commented:
So I made the changes I said I was going to make. Checking for special interior now on raise water level, removing 'entrance' tag from goron city grotto, adding 'skulltulas' to tag to spirit temple ledge jump, and adding a bit to the tooltip for the shadow falling spikes gs to recommend enabling the equivalent trick to obtain the chests.

r0bd0g commented:
Some location names got changed since I put this in. Obviously don't accidentally change the names back in the OW logic.

Zannick commented on SettingsList.py:
```diff
@@ -891,6 +949,18 @@ def __init__(self, name, gui_text, min, max, default, step=1,
                     in the Water Temple are not going to be relevant unless this
                     trick is first enabled.
                     '''},
+    'Water Temple Central Pillar GS with Neither Longshot nor Farore\'s Wind': {
```
Why not "... with Iron Boots"?

r0bd0g replied:
That could be a better name. This name is pretty wordy. I don't know how I feel about "with Iron Boots" because a lot of the time the entire water temple is just requiring irons boots, so to say "with Iron Boots" about anything in Water Temple seems self-evident until you realize it's referring to a specific trick that you need the iron boots to do. Does that make any sense? But I've called tricks in the water temple "with Iron Boots" before, so I'm definitely not opposed to doing it.

Part of why I chose to do it like this is I was setting up for potentially turning the FW route into a trick (which I'm definitely leaning in favour of doing), and I wanted the names of the tricks to contrast. So I was thinking "without Longshot using Farore's Wind" (also very wordy) for the new FW trick. So regardless you do have to keep both the Iron Boots and Farore's Wind tricks in mind when deciding names.

Zannick replied:
I think trick names will get pretty wordy if you list both the items you avoid and the items you need. I would just call it "with Farore's Wind" and "with Iron Boots" or whatever. Looking for examples, there's "Dodongo's Cavern Staircase With Bow" doesn't say "without Strength or Bombs using Bow". We don't have to be exact in the name, just make it easy to distinguish what trick a name refers to, and maybe be concise at the same time. So I like it more when we can specify "with Item" or "without Item" to uniquely refer to a trick.

r0bd0g replied:
At one point I was thinking of having just one trick, "without Longshot" encompass all of it. So in my mind they were both ways of skipping Longshot, so that's why I ended up with "without Longshot" in the name.

Zannick reviewed (changes requested):
Looks good, mainly typos, questions, suggestions.

Regarding Farore's Wind:

Water Temple GS is one of the most complicated locations to get. After this PR, you can get it with

- Longshot (1st or 2nd entrance)
- FW + Hookshot + (2nd floor entrance: fire)
- child + FW + Boomerang + ((2nd floor entrance: fire) or (1st floor entrance: keys + (as adult, hovers or bow)))
- the no_fw trick + Iron Boots + Hookshot + (2nd floor entrance: fire)

I don't know that I would want FW to be foolish ever. Currently it's not, if I've read the code right (major items have to be explicitly excluded, or a duplicate of a WoTH item, to be foolish). But I agree, it's definitely weird for FW to ever be actually WoTH because it's literally only this location where it's needed. So I'm okay with putting both the FW cases behind a FW trick for this location. That way, with tricks off, the only logical way to get it is with longshot. And I think I like the iron boots trick more than requiring FW for this skull.

r0bd0g commented on SettingsList.py:
```diff
@@ -747,6 +791,13 @@ def __init__(self, name, gui_text, min, max, default, step=1,
+        'name'    : 'logic_ice_block_gs',
+        'tags'    : ("Ice Cavern", "Skulltulas",),
+        'tooltip' : '''\
+                    The Hover Boots can be used to kill the Skulltula
```
I wanted to make it clear that the Hover Boots can be used to get in front of the skulltula in order to jumpslash it. I'll be a bit more wordy here and spell that out.

r0bd0g commented on SettingsList.py:
```diff
@@ -983,6 +1079,15 @@ def __init__(self, name, gui_text, min, max, default, step=1,
+        'tooltip' : '''\
+                    With help from a conveniently-positioned block,
+                    Adult can quickly carry a bomb flower over to
+                    destroy the mud wall blockng the room with two
```
My keyboard kind of sucks and sometimes drops characters, particularly the i key can be an issue. I struggled so hard with typing all of this TBH and I'm not surprised some made it through unnoticed.

r0bd0g commented on data/World/Forest Temple.json:
```diff
@@ -5,7 +5,10 @@
+            "Forest Temple GS First Room": "
+                (is_adult and (Hookshot or Bow or Bombs)) or (is_child and (Boomerang or Slingshot)) or
+                has_bombchus or can_use(Dins_Fire) or (logic_forest_first_gs and (Bombs or
+                    ((is_adult or Sticks or Kokiri_Sword) and (damage_multiplier != 'ohko' or Fairy or can_use(Nayrus_Love)))))",
```
Thanks, I'll take a look at it.

r0bd0g commented on data/World/Forest Temple.json:
```diff
@@ -30,6 +33,9 @@
         "exits": {
+            "Forest Temple NE Outdoors": "
+                can_use(Iron_Boots) or can_use(Longshot) or 
+                (Progressive_Scale, 2) or (logic_forest_well_swim and can_use(Hookshot))",
```
It was brought up that having gold scale as child could skip needing an item for child to kill the bubble in order to reach the other courtyard, but the only thing for child to do over there is the boomerang trick for the gs. With boomerang you can kill babas and get the necessary weapons for the bubble anyway. So all this entrance does is give the playthrough more options as to which items it decides to pick up for getting to that gs as child. It's practically irrelevant as child and actually does nothing at all as adult. I decided to just put the full logic in. The tooltip is not wrong about the well swim trick only possibly being required in MQ. That's still true.

Come to think of it, this change is not even complete b/c the bubble kill is permanent, and in theory you could kill the bubble as adult so that child can get through and use the boomerang on the skulltula as child. So the entrance from NW to High Balconies would have to be wrapped in a here(). Thinking about it some more I'm thinking this change is probably not worth making. Should I just do away with this well exit? All it's for it so the playthrough doesn't pick up unnecessary items.

Zannick replied:
The name on the trick could be misleading, if it alters logic that could affect gameplay. If it would only affect the playthrough picking up "unnecessary" items, I don't think it's worth it.

r0bd0g replied:
What's not worth it?
1. We can not make any changes, and unnecessary items get picked up in the playthrough.
2. (Requires adding a here() to the high balconies exit, which I missed before.) We can make it so playthrough won't pick up unnecessary items ...
2a) by having the full logic on the well connection.
2b) by only checking for gold scale on the well connection.

r0bd0g commented on data/World/Shadow Temple.json:
```diff
@@ -42,14 +42,15 @@
                     (logic_shadow_freestanding_key and has_bombchus))",
             "Shadow Temple GS Like Like Room": "True",
-            "Shadow Temple GS Falling Spikes Room": "Progressive_Hookshot",
+            "Shadow Temple GS Falling Spikes Room": "logic_shadow_umbrella_gs or Progressive_Hookshot",
```
In vanilla Shadow, you need hovers to get past the truth spinner and onto the "tongue". In MQ, there's a platform that you can spawn with Fire Arrows, so in MQ it's possible to be deeper in without Hovers.

r0bd0g commented on data/World/Water Temple.json:
```diff
@@ -114,39 +121,42 @@
+                (logic_water_falling_platform_gs_boomerang and can_use(Boomerang))"
+        },
+        "exits": {
+            "Water Temple Dark Link Region": "(Small_Key_Water_Temple, 6) and can_use(Hookshot)"
```
The falling platform room used to be in dark link region, but I had to separate it off when giving child access in order to boomerang the token. The exit to that room is still 5 keys. Exiting from that room to the rest of the dark link region there was no reason to keep the 6 keys checks on the locations themselves, since now there's a region entirely behind the 6-key door.

r0bd0g commented:
Alright, changes:
1. Reworded that ice cavern hovers tooltip to make it clearer how the hovers are used to obtain this token.
2. i
3. can_jumpslash
4. Reduced the forest well crossing from West to East to just the logically-relevant (only for the playthrough) Gold Scale and made a comment pointing that out as it's possible that in the future other methods could become relevant. I maybe should have said "in this direction" so that people reading it don't think that this exit is for passing through the well the other way, where plenty more options are logically important. I also added a here() around the access from west courtyard to high balconies to capture adult defeating the bubble for child to pass through. Reworded the logic on that pass through a bit.
5. Renamed that trick to "with Iron Boots".
